### PR TITLE
Suggest user to use WSL in Windows instructions

### PR DESCRIPTION
**Pull request by ktbarrett**
Addresses some of #2843. This follows what we decided in the last meeting.

The second part of the instructions which recommend Conda + mingw compiler will be replaced with python.org installer once we have wheels and are happy with the python runner.

### CODE REVIEW

File: documentation/source/install.rst [hunk 1]
```diff
@@ -34,8 +34,11 @@ The installation instructions vary depending on your operating system:
 
    .. group-tab:: Windows
 
-      `Conda <https://conda.io/>`_ is an open-source package and environment management system that we recommend for Windows.
+      We recommend users who are comfortable with a Unix shell and who are running Windows 10 or 11 use Windows Subsytem for Linux (WSL).
+      See `here <https://docs.microsoft.com/en-us/windows/wsl/install>`_ for installation instructions of WSL,
+      then follow the Linux installation instructions for cocotb.
```
Comment (marlonjames) [inline documentation/source/install.rst:39]:
```suggestion
      We recommend users who are comfortable with a Unix shell and who are running Windows 10 or 11 use Windows Subsystem for Linux (WSL).
      See `here <https://docs.microsoft.com/en-us/windows/wsl/install>`_ instructions on installing WSL,
      then follow the Linux installation instructions for cocotb.
```

**Review by marlonjames — APPROVED** [review]

File: documentation/source/install.rst (hunk 1 above)
Comment (marlonjames) [reply documentation/source/install.rst:39]: hah, yeah

**Comment by ktbarrett** [comment]
> What's the simulator story there?

I presume you are concerned about GUI features? There are two possibilities. Windows 11 has WSLg, which is builtin GUI support with WSL, but it's pretty bad compared to the alternative right now. Alternatives are X servers for Windows, of which there are a couple. I use VcXsrv. That works flawlessly from what I can tell.

**Comment by marlonjames** [comment]
I use WSL on Windows 10 for iverilog, GHDL, and Verilator. But those are non-GUI.

I have successfully used Aldec Active-HDL with both system Python and conda on Windows 7 and 10.

**Comment by ktbarrett** [comment]
The thought is on Windows to recommend one of 2 solutions.
1. Use WSL is you are comfortable with a Unix shell *instead of* mingw, msys, cygwin, etc, etc, etc.
2. Use Anaconda + mingw/msvc if you are comfortable with Windows.

Right now we have no Tier 1 supported recommendation for point 1 since most alternative Unix subsystems are broken or hokey. This PR is adding an official recommendation for point 1 to use WSL, implying that other Unix subsystems are not recommended. The recommendation for point 2 will stay the same for now. The recommendation for point 2 will change in the future to system Python installation + pre-built wheel, so no need for a compiler + Python runner, so no need for make. The anaconda solution will be relegated to Tier 2 and installation instructions moved to the Wiki.

The end result is something that Windows people will love and Linux people who are forced to use Windows will love.

**Comment by imphil** [comment]
I have a hard time seeing the user base for WSL: Windows users who are familiar with Linux, don't want to use a proprietary simulator, and don't want to use a full Linux in a VirtualBox or HyperV VM?

Or in other words: Given the support matrix for proprietary simulators I don't think "being comfortable with Unix: yes/no" is the right decision criteria. It's more: do you want to use a proprietary simulator: yes/no? And then the rest falls into place.

Looking at our user survey two years ago, the majority of Windows users are using Aldec or Mentor simulators. 

Given the really limited amount of resources we have available for Windows support: How about just dropping WSL and providing a single Tier 1 solution that works on Windows together with proprietary simulators? And I agree, system python + wheels seems right now the best solution, and is very close to becoming a reality.

And if users want an easy Linux "apt-get install icarus gtkwave; pip install cocotb; make my_test; gtkwave *.fst" experience, just use a Ubuntu VM.

**Comment by ktbarrett** [comment]
> Given the support matrix for proprietary simulators

I'm not sure where you are getting this from? They all work on WSL. GUI stuff requires extra effort, but it's not difficult, or really our responsibility. There are a million medium articles on setting up an X server for WSL. There are also advantages of WSL that are not available to a VM. I haven't tried it, but it is theoretically possible to program a board from WSL.

> Given the really limited amount of resources we have available for Windows support: How about just dropping WSL

WSL requires no additional support from us.

I think you are missing the point. We need to recommend _something_ for users who are forced to use Windows (common) and would prefer a Unix environment (possible) or have legacy makefile-based projects (possible) or they will pick the _something_ and inevitably come bothering us about it when it doesn't work. Recommending WSL is a solution, recommending a VM is also a solution, but a less ideal one for people doing EDA IMO.

**Comment by gretzteam** [comment]
I have been using cocotb on wsl with xcelium for about 2 years. I beleive this is the best setup for windows users. Setting up the gui is trivial which is only really needed for the waveform viewer. Everything else is done through the fantastic vscode with wsl extension.
This setup allows you to use windows for day to day task (if you click on a xls file it opens it in excel, not libreoffice.😉). At the same time you have a full blown Linux dev machine.
I highly recommend it.

**Comment by imphil** [comment]
> I have been using cocotb on wsl with xcelium for about 2 years.

And I assume you're using the Linux version of Xcelium installed into WSL?

**Comment by gretzteam** [comment]
> > I have been using cocotb on wsl with xcelium for about 2 years.
> 
> And I assume you're using the Linux version of Xcelium installed into WSL?

Yes. I would actually do the same for any simulator as it's just easier to install in Linux. Verilator, icarus...even python. 
Then Linux stuff runs on Linux and you use windows for interacting with the computer. Where you draw the line is up to the user.

**Comment by ktbarrett** [comment]
Another justification for recommending something over not recommending anything https://gitter.im/cocotb/Lobby?at=621cb0ffc61ef0178e4df216.

**Review by cmarqu — APPROVED** [review]

File: documentation/source/install.rst [hunk 2]
```diff
@@ -34,8 +34,11 @@ The installation instructions vary depending on your operating system:
 
    .. group-tab:: Windows
 
-      `Conda <https://conda.io/>`_ is an open-source package and environment management system that we recommend for Windows.
+      We recommend users who are comfortable with a Unix shell and who are running Windows 10 or 11 use Windows Subsystem for Linux (WSL).
+      See `here <https://docs.microsoft.com/en-us/windows/wsl/install>`_ for instructions on installing WSL,
+      then follow the Linux installation instructions for cocotb.
```
Comment (ktbarrett) [inline documentation/source/install.rst:39]: @imphil Is this any better?

```suggestion
      We recommend users who are running Windows, and who are more comfortable with a Unix shell or who have legacy Makefile-based projects
      to use either Windows Subsystem for Linux (WSL) or a Virtual Machine, such as Oracle VirtualBox.
      After installing WSL or a Virtual Machine and a supported Linux distribution, follow the Linux installation instructions for cocotb.
```

File: documentation/source/install.rst [hunk 3]
```diff
@@ -34,8 +34,11 @@ The installation instructions vary depending on your operating system:
...
 
+      `Conda <https://conda.io/>`_ is an open-source package and environment management system that we recommend for users who are more comfortable with Windows.
```
Comment (ktbarrett) [inline documentation/source/install.rst:41]:
```suggestion
      `Conda <https://conda.io/>`_ is an open-source package and environment management system that we recommend for users who are more comfortable with native Windows development.
```

**Comment by umarcor** [comment]
For GUIs on Windows (either on WSL or using containers) check [x11docker](https://github.com/mviereck/x11docker/) and [runx](https://github.com/mviereck/runx). See https://hdl.github.io/containers/ug/GUI.html.

Alternatively, as explained in https://github.com/ghdl/docker/blob/master/USE_CASES.md#environments-with-guis GTKWave can use GTK3's Broadway feature and be used through a browser. Container `ghdl/ext:broadway` includes the following script: https://github.com/ghdl/docker/blob/master/broadway.sh. Note that some builds of GTKWave use GTK2; this needs GTK3.

VSCode's Docker tab allows automating the options to start containers with x11docker or with predefined ports bind to the host.

**Comment by umarcor** [comment]
> it is theoretically possible to program a board from WSL since you have direct access to Windows devices.

@ktbarrett that works for UART/COM and disk drives only, not for FPGAs (libusb, FTDI, etc.). In order to program FPGAs from a container (say HyperV, say WSL2), USB/IP is required. See https://hdl.github.io/containers/ug/USBIP.html.

File: documentation/source/install.rst [hunk 4]
```diff
@@ -34,8 +34,11 @@ The installation instructions vary depending on your operating system:
 
    .. group-tab:: Windows
 
-      `Conda <https://conda.io/>`_ is an open-source package and environment management system that we recommend for Windows.
+      We recommend users who are running Windows, and who are more comfortable with a Unix shell or who have legacy Makefile-based projects,
+      to use either Windows Subsystem for Linux (WSL) or a Virtual Machine, such as Oracle VirtualBox.
```
Comment (ktbarrett) [inline documentation/source/install.rst:38]:
```suggestion
      We recommend users who are running Windows and who are more comfortable with a Unix shell or who have legacy Makefile-based projects to use either Windows Subsystem for Linux (WSL) or a Virtual Machine such as Oracle VirtualBox.
```
Better grammar I think? @cmarqu

**Comment by ktbarrett** [comment]
@umarcor What are the differences between using runx or x11docker and running an X11 server and configuring your `DISPLAY`?

> that works for UART/COM and disk drives only

That's a shame.

**Comment by ktbarrett** [comment]
Should we point to a Wiki page which contains more in-depth living documentation on setting up WSL2 for cocotb's use?

**Comment by umarcor** [comment]
> What are the differences between using runx or x11docker and running an X11 server and configuring your `DISPLAY`?

x11docker is a 10kLOC bash script to automate running the server and passing ports/sockets/files to the container. It's essentially the same. However, x11docker includes multiple security and performance related features. It applies sane options to prevent the server being accessed by any other than the intended container, it supports xpra/wayland/others, window or desktop mode, etc. I strongly recommend to read the two page paper and the wiki.

Disclaimer: I triggered the author to support windows and I helped him test with MSYS2, cygwin, WSL, xming, vcxsrv... Then, runx was split from x11docker, which I also helped test.

File: documentation/source/install.rst (hunk 4 above)
Comment (cmarqu) [reply documentation/source/install.rst:38]: Uh, not sure really... based on feeling I might even add a comma before the "or who have".

File: documentation/source/install.rst [hunk 5]
```diff
@@ -34,8 +34,12 @@ The installation instructions vary depending on your operating system:
 
    .. group-tab:: Windows
 
-      `Conda <https://conda.io/>`_ is an open-source package and environment management system that we recommend for Windows.
+      We recommend users who are running Windows and who are more comfortable with a Unix shell,
+      or who have legacy Makefile-based projects,
+      to use Windows Subsystem for Linux (WSL).
+      After installing WSL or a Virtual Machine and a supported Linux distribution, follow the Linux installation instructions for cocotb.
```
Comment (ktbarrett) [inline documentation/source/install.rst:40]:
```suggestion
      After installing WSL and a supported Linux distribution, follow the Linux installation instructions for cocotb.
```

**Pull request merged**